src/stubs/sm_output.c

```diff
@@ -5,11 +5,26 @@
 SM_FUNC(SM_NAME) void __sm_send_output(io_index index,
                                        const void* data, size_t len)
 {
-    const size_t nonce_len = sizeof(__sm_output_nonce);
-    const size_t payload_len = nonce_len + len + SANCUS_TAG_SIZE;
-    uint8_t* payload = malloc(payload_len);
-    *(uint16_t*)payload = __sm_output_nonce++;
-    sancus_wrap_with_key(__sm_io_keys[index], payload, nonce_len, data, len,
-                         payload + nonce_len, payload + nonce_len + len);
-    reactive_handle_output(index, payload, payload_len);
+    const size_t payload_len = len + SANCUS_TAG_SIZE;
+
+    // search for all the connections associated to the index.
+    // Unfortunately, this operation is O(n) with n = number of connections
+    int i;
+    for (i=0; i<__sm_num_connections; i++) {
+      Connection *conn = &__sm_io_connections[i];
+      if(conn->io_id != index)
+        continue;
+
+      uint8_t* payload = malloc(payload_len);
```

Comment:
this is an _untrusted_ `malloc` call to allocate memory outside the enclave right?

--> You should check the resulting `payload` pointer and length lies outside the enclave! Otherwise `sancus_wrap_with_key` may be tricked in overwriting in-enclave memory :smiling_imp:

Reply:
damn it! you're right as usual 😀